Add production checklist for ScalarDL (#215)
scalar-labs/scalar-kubernetes: docs/ProductionChecklistForScalarDLLedger.md

### Network

You should create the Kubernetes cluster on a private network since ScalarDL Ledger does not provide any services to users directly via internet access. We recommend accessing ScalarDL Ledger via a private network from your applications.

### Monitoring and logging

You should monitor the deployed components and collect their logs. For details, see [Monitoring Scalar products on a Kubernetes cluster](./K8sMonitorGuide.md) and [Collecting logs from Scalar products on a Kubernetes cluster](./K8sLogCollectionGuide.md).

### Backup and restore

You should enable the automatic backup feature and point-in-time recovery (PITR) feature in the backend database. For details, see [Set up a database for ScalarDB/ScalarDL deployment](./SetupDatabase.md).

## Production checklist: Client applications that access ScalarDL Ledger

The following is a checklist of recommendations when setting up a client application that accesses ScalarDL Ledger in a production environment.

### Do not deploy your application on the same Kubernetes cluster as ScalarDL Ledger

> For Byzantine fault detection in ScalarDL to work properly, do not deploy your application pods on the same Kubernetes clusters as the ScalarDL Ledger deployment. Please deploy your application on an environment other than the administrative domain (other than the Kubernetes cluster) for ScalarDL Ledger deployment.

Reviewer: ```suggestion
For Byzantine fault detection in ScalarDL to work properly, do not deploy your application pods on the same Kubernetes clusters as the ScalarDL Ledger deployment. Instead, you must deploy your application in an environment other than the administrative domain (other than the Kubernetes cluster) for the ScalarDL Ledger deployment.
```

#### Required for production environments

```mermaid
graph LR
  subgraph "Administrative domain 1"
    subgraph "Another environment"
      A-1[User application]
    end
  end
  subgraph "Administrative domain 2"
    subgraph "Kubernetes cluster"
      B-1[ScalarDL Ledger]
    end
  end
  A-1 --> B-1
```

#### Not recommended for production environments (for testing purposes only)

```mermaid
graph LR
  subgraph "Kubernetes cluster"
    direction LR
    A-1[User application] --> A-2[ScalarDL Ledger]
  end
```

### Check if your contract and function follow guidelines

> You should check if your contract and function follow the guidelines:

Reviewer: ```suggestion
To check if your contract and function follow the guidelines, see the following:
```

* [A Guide on How to Write a Good Contract for ScalarDL](https://github.com/scalar-labs/scalardl/blob/master/docs/how-to-write-contract.md)
* [A Guide on How to Write Function for ScalarDL](https://github.com/scalar-labs/scalardl/blob/master/docs/how-to-write-function.md)

### Versioning contracts

> When you register a contract once, you cannot overwrite the existing contract. So, you should consider the versioning of contracts. We recommend one of the following two methods:

Reviewer: ```suggestion
After you register a contract, you cannot overwrite that existing contract. So, you should consider the versioning of contracts. We recommend one of the following two methods.
```

#### Versioning by using `Class Name`

```console
Contract ID              : FooV1
Binary Name              : com.example.contract.FooV1
Class file (Class Name)  : src/main/java/com/example/contract/FooV1.class
---
Contract ID              : FooV2
Binary Name              : com.example.contract.FooV2
Class file (Class Name)  : src/main/java/com/example/contract/FooV2.class
```

#### Versioning by using `Package Name`

```console
Contract ID             : FooV3
Binary Name             : com.example.contract.v3.Foo
Class file (Class Name) : src/main/java/com/example/contract/v3/Foo.class
---
Contract ID             : FooV4
Binary Name             : com.example.contract.v4.Foo
Class file (Class Name) : src/main/java/com/example/contract/v4/Foo.class
```

### Use the same binary name, package name, and class name when you register the contract

> If the binary name, package name, and class name are different when you register the contract, you cannot execute such a contract after you register it.

Reviewer: ```suggestion
If the binary name, package name, and class name are different when you register the contract, you cannot execute that contract after registering it.
```

#### Binary name and class name are different (you cannot execute this contract)

```console
Contract ID              : FooV5
Binary Name              : com.example.contract.FooV5
Class file (Class Name)  : src/main/java/com/example/contract/FooV6.class
```

#### Binary name and package name are different (you cannot execute this contract)

```console
Contract ID              : FooV7
Binary Name              : com.example.contract.v7.Foo
Class file (Class Name)  : src/main/java/com/example/contract/v8/Foo.class
```

> ### Secret key and certification file requirements

Reviewer: ```suggestion
### Secret key and certificate file requirements
```

> You must make sure that your secret key and certification file match the following ScalarDL requirements.

Reviewer: ```suggestion
You must make sure that your secret key and certificate file match the following ScalarDL requirements.
```

> You must make sure that your secret key and certification file match the following ScalarDL requirements.

Reviewer 2: The comment [here](https://github.com/scalar-labs/scalar-kubernetes/pull/215#discussion_r1375703097) should also be applied to the above?

Author: Ah, yes. You are right.
I overlooked it...
Thank you for pointing it out!
I fixed it in d2857e8c9001f36ea0d57ef4e89ab3dfda2082b6.

```console
Algorithm       : ECDSA
Hash function   : SHA256
Curve parameter : P-256
```

For details, see [How to get a certificate](https://github.com/scalar-labs/scalardl/blob/master/docs/ca/caclient-getting-started.md).

### Exception handling

> You must make sure that your application handles exceptions. For details, see [A Guide on How to Handle Errors in ScalarDL](https://github.com/scalar-labs/scalardl/blob/master/docs/how-to-handle-errors.md).

Reviewer: ```suggestion
You must make sure that your application handles exceptions. For details, see [A Guide on How to Handle Errors in ScalarDL](https://github.com/scalar-labs/scalardl/blob/master/docs/how-to-handle-errors.md).

```

## Before you begin

In this checklist, we assume that you are deploying ScalarDL Ledger on a managed Kubernetes cluster, which is recommended.

## Production checklist: ScalarDL Ledger

The following is a checklist of recommendations when setting up ScalarDL Ledger in a production environment.

### Number of pods and Kubernetes worker nodes

To ensure that the Kubernetes cluster has high availability, you should use at least three worker nodes and deploy at least three pods spread across the worker nodes. You can see the [sample configurations](../conf/scalardl-custom-values.yaml) of `podAntiAffinity` for making three pods spread across the worker nodes.

{% capture notice--info %}
**Note**

If you place the worker nodes in different availability zones (AZs), you can withstand an AZ failure.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Worker node specifications

From the perspective of commercial licenses, resources for one pod running ScalarDL Ledger are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDL Ledger pods exist on the worker nodes.

> In other words, the following components could run on one worker node:

Reviewer: Does the wording for this depend on how we word a similar sentence in a different PR?: [related comment](https://github.com/scalar-labs/scalar-kubernetes/pull/213/files#r1382880277)
